syntax = "proto3";

service NanoModal {
    rpc Invoke(InvokeRequest) returns (InvokeResponse);
    rpc GetResult(GetResultRequest) returns (GetResultResponse);
    rpc InvokeMany(InvokeManyRequest) returns (InvokeManyResponse);
    rpc InvokeManyStream(InvokeManyRequest) returns (stream StreamResult);
}

message InvokeRequest {
    bytes function_pickle = 1;
    bytes args_pickle = 2;
}

message InvokeResponse {
    string task_id = 1;
}

message InvokeManyRequest {
    bytes function_pickle = 1;
    repeated bytes args_pickles = 2;
}

message InvokeManyResponse {
    repeated string task_ids = 1;
}

message StreamResult {
    int32 index = 1;
    bytes result_pickle = 2;
    string error = 3;
}

message GetResultRequest {
    string task_id = 1;
}

message GetResultResponse {
    bytes result_pickle = 1;
    string error = 2;
}